### PR TITLE
Use official Android Gradle plugin

### PR DESCRIPTION
**Pull request by lazar-mitrovic**
This PR:
* Changes `package` phase, so that it generates Gradle Android project using official plugin in `/target/client/aarch64-android/gvm/android_project` and delegates apk building/signing/deployment to it. This results in much faster package/install phases, enables usage of android specific libraries and enables importing in Android Studio.

* Sets ARM64 backend for Android building as default which should significantly speed up compilation, and use less peak memory.

### CODE REVIEW

**Comment by eugener** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gluonhq/substrate/pullRequest?prid=5574548&bid=18275932)

File: src/main/java/com/gluonhq/substrate/util/FileOps.java [hunk 1]
```diff
@@ -141,6 +148,44 @@ public static Path copyResource(String resource, Path destination) throws IOExce
...
+            @Override
+            public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+                currentTarget = target.resolve(jarPath.relativize(dir).toString());
+                Files.createDirectories(currentTarget);
+                return FileVisitResult.CONTINUE;
+            }
+
+            @Override
+            public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                Files.copy(file, target.resolve(jarPath.relativize(file).toString()), StandardCopyOption.REPLACE_EXISTING);
```
Comment (eugener) [inline src/main/java/com/gluonhq/substrate/util/FileOps.java:182]: Codacy Issue found: [Unnecessary use of fully qualified name 'StandardCopyOption.REPLACE_EXISTING' due to existing static import 'java.nio.file.StandardCopyOption.REPLACE_EXISTING'](https://app.codacy.com/manual/gluonhq/substrate/pullRequest?prid=5574548)

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java [hunk 2]
```diff
@@ -221,22 +173,33 @@ public boolean runUntilEnd() throws IOException, InterruptedException {
         return processResult == 0;
     }
 
+    boolean useLLVM = false;
```
Comment (eugener) [inline src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:176]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/gluonhq/substrate/pullRequest?prid=5574548)

File: src/main/java/com/gluonhq/substrate/util/FileOps.java [hunk 3]
```diff
@@ -141,6 +148,44 @@ public static Path copyResource(String resource, Path destination) throws IOExce
...
+            fileSystem = FileSystems.newFileSystem(resource, Collections.<String, String>emptyMap());
+        } catch(URISyntaxException e) {
+            throw new IOException(e.toString());
+        }
+
+        final Path jarPath = fileSystem.getPath(source);
+
+        Files.walkFileTree(jarPath, new SimpleFileVisitor<Path>() {
+
+            private Path currentTarget;
```
Comment (eugener) [inline src/main/java/com/gluonhq/substrate/util/FileOps.java:171]: Codacy Issue found: [Perhaps 'currentTarget' could be replaced by a local variable.](https://app.codacy.com/manual/gluonhq/substrate/pullRequest?prid=5574548)

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java [hunk 4]
```diff
@@ -124,80 +106,50 @@ public boolean compile() throws IOException, InterruptedException {
...
-        }
+        prepareAndroidProject();
+        prepareAndroidManifest();
+        prepareAndroidResources();
+        copyOtherDalvikClasses();
+        copySubstrateLibraries();
 
-        processResult = sign(buildToolsPath, alignedApk);
-        return processResult == 0;
+        ProcessRunner assembleDebug = new ProcessRunner(
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:123]: I take that `assembleDebug` will sign internally the apk for debug, and the `debug.keystore` we created is no longer required, right?
What about signing for release? I see that you have removed the `sign` method. What is the alternative? Could we create the signing configuration in the build.gradle file, with the release data?

File: src/main/java/com/gluonhq/substrate/util/FileOps.java [hunk 5]
```diff
@@ -141,6 +148,44 @@ public static Path copyResource(String resource, Path destination) throws IOExce
...
+        // https://stackoverflow.com/a/24316335
+        FileSystem fileSystem;
+        try {
+            URI resource = SubstrateDispatcher.class.getResource("").toURI();
+            fileSystem = FileSystems.newFileSystem(resource, Collections.<String, String>emptyMap());
+        } catch(URISyntaxException e) {
+            throw new IOException(e.toString());
+        }
+
+        final Path jarPath = fileSystem.getPath(source);
```
[the remark this answers is not in the document]
Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/util/FileOps.java:167]: I believe that calling `FileOps::copyDirectory` in `preVisitDirectory` would copy all directories several times (for every node depth)? 
If that is what you meant?

[the remark this answers is not in the document]
Comment (jperedadnr) [reply src/main/java/com/gluonhq/substrate/util/FileOps.java:167]: I meant that we already have `copyDirectory(Path, Path)`, so you could replace the whole `Files.walkFileTree(...)` part with it. I haven't tested it though.

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java [hunk 6]
```diff
@@ -407,122 +300,27 @@ private void copyOtherDalvikClasses() throws IOException, InterruptedException {
...
-        Path substrateLibPath = getApkLibArm64Path().resolve("libsubstrate.so");
-        Files.copy(libPath, substrateLibPath, StandardCopyOption.REPLACE_EXISTING);
-
-        List<String> aaptAddLibsArgs = new ArrayList<>(Arrays.asList(aaptCmd, "add", unalignedApk, "lib/arm64-v8a/libsubstrate.so"));
-
+    /*
+     * Copies native libraries to android project
+     */
+    private void copySubstrateLibraries() throws IOException {
+        Path projectLibsLocation = getAndroidProjectMainPath().resolve("jniLibs").resolve("arm64-v8a");
```
[the remark this answers is not in the document]
Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:319]: Fixed

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java [hunk 7]
```diff
@@ -300,94 +263,24 @@ boolean useGraalVMJavaStaticLibraries() {
...
      * and looks for META-INF/substrate/dalvik/*.class files.
      *
-     * The method will copy all the class files found into the target folder
+     * The method will copy all the class files found into jar in the target folder
      *
      * @throws IOException
      */
     private void copyOtherDalvikClasses() throws IOException, InterruptedException {
-        Path targetFolder = getApkClassesPath();
+        Path targetFolder = getAndroidProjectPath().resolve("libs").resolve("tmp");
```
[the remark this answers is not in the document]
Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:282]: Fixed

**Comment by eugener** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gluonhq/substrate/pullRequest?prid=5574548&bid=18275932)

File: src/main/java/com/gluonhq/substrate/util/FileOps.java [hunk 8]
```diff
@@ -48,23 +48,29 @@
...
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
 import java.nio.file.FileVisitOption;
 import java.nio.file.FileVisitResult;
 import java.nio.file.FileVisitor;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.nio.file.SimpleFileVisitor;
+import java.nio.file.StandardCopyOption;
```
Comment (eugener) [inline src/main/java/com/gluonhq/substrate/util/FileOps.java:66]: Codacy Issue found: [Unused import - java.nio.file.StandardCopyOption.](https://app.codacy.com/manual/gluonhq/substrate/pullRequest?prid=5574548)

Comment (eugener) [inline src/main/java/com/gluonhq/substrate/util/FileOps.java:66]: Codacy Issue found: [Avoid unused imports such as 'java.nio.file.StandardCopyOption'](https://app.codacy.com/manual/gluonhq/substrate/pullRequest?prid=5574548)

File: src/main/resources/native/android/android_project/gradle/wrapper/gradle-wrapper.properties [hunk 9]
```diff
@@ -0,0 +1,5 @@
+distributionBase=GRADLE_USER_HOME
+distributionPath=wrapper/dists
+zipStoreBase=GRADLE_USER_HOME
+zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.6.4-all.zip
```
[the remark this answers is not in the document]
Comment (lazar-mitrovic) [reply src/main/resources/native/android/android_project/gradle/wrapper/gradle-wrapper.properties]: That was combination of android plugin and gradle that I found worked during thinkering.
Latest commit bumped everything up to newest versions.

File: src/main/java/com/gluonhq/substrate/util/FileOps.java [hunk 10]
```diff
@@ -264,6 +309,28 @@ public static void copyDirectory(Path source, Path destination) throws IOExcepti
...
 
+    /**
+     * Recursively list files with spectified extension from directory
+     * @param directory directory to be searched
+     * @param extension extension by which to filter files
+     * @throws IOException if an exception happens when listing the content
+     */
+    public static List<String> listFilesWithExtensionInDirectory(Path directory, String extension) throws IOException {
+        try (Stream<Path> walk = Files.walk(directory)) {
+            return walk.filter(Files::isRegularFile).map(x -> x.toString())
```
[the remark this answers is not in the document]
Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/util/FileOps.java]: Fixed

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java [hunk 11]
```diff
@@ -221,22 +173,33 @@ public boolean runUntilEnd() throws IOException, InterruptedException {
...
-        return Arrays.asList("-H:CompilerBackend=" + Constants.BACKEND_LLVM,
+        ArrayList<String> flags = new ArrayList<String>(Arrays.asList(
                 "-H:-SpawnIsolates",
                 "-Dsvm.targetArch=" + projectConfiguration.getTargetTriplet().getArch(),
                 "-H:+UseOnlyWritableBootImageHeap",
                 "-H:+UseCAPCache",
                 "-H:CAPCacheDir=" + getCapCacheDir().toAbsolutePath().toString(),
-                "-H:CustomLD=" + ldlld.toAbsolutePath().toString());
+                "-H:+UseBionicC"));
+        if (useLLVM) {
```
[the remark this answers is not in the document]
Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java]: Fixed

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java [hunk 12]
```diff
@@ -407,122 +300,27 @@ private void copyOtherDalvikClasses() throws IOException, InterruptedException {
...
-            return processResult;
-        }
-
-        processResult = aaptAddDxClasses(buildToolsPath, unalignedApk);
-        if (processResult != 0) {
-            return processResult;
-        }
+        ProcessRunner createJar = new ProcessRunner(
+                projectConfiguration.getGraalPath().resolve("bin").resolve("jar").toString(),
+                "-cvf", "../attach.jar", ".");
```
[the remark this answers is not in the document]
Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java]: I renamed it to `additional_classes.jar`, if you have better idea let me know :)

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java [hunk 13]
```diff
@@ -532,86 +330,58 @@ private int sign(Path buildToolsPath, String alignedApk) throws IOException, Int
...
 
     /**
-     * If android manifest is present in src/android, then, this
-     * path will be returned.
+     * If android manifest is present in src/android, it will be copied to
+     * android project.
      *
-     * Else, default android manifest is copied into gvm/genSrc/android and
-     * this path is returned
+     * Else, default android manifest is adjusted and used in project
```
[the remark this answers is not in the document]
Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:364]: I agree, I moved `android_project` directory to `gensrc` since it seems more appropriate

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java (hunk 4 above)
Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:128]: This should be fixed in latest commit

File: src/main/java/com/gluonhq/substrate/util/FileOps.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/util/FileOps.java:166]: This fails with `java.nio.file.ProviderMismatchException`

File: src/main/java/com/gluonhq/substrate/util/FileDeps.java [hunk 14]
```diff
@@ -75,7 +75,7 @@
 
     private static final String ANDROID_KEY = "24333f8a63b6825ea9c5514f83c2829b004d1fee";
     private static final String[] ANDROID_SDK_PACKAGES = {
-            "platforms;android-27", "build-tools;27.0.3", "platform-tools", 
+            "platforms;android-27", "build-tools;29.0.2", "platform-tools",
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/util/FileDeps.java:78]: I get this warning at the end of the package task:
```
WARNING: No version of NDK matched the required version 21.1.6352462. Versions available locally: 
```
I have installed the ndk-bundle, but somehow it doesn't find it. Not sure if this is an issue?

File: src/main/resources/native/android/android_project/app/build.gradle [hunk 15]
```diff
@@ -2,15 +2,14 @@ apply plugin: 'com.android.application'
 
 android {
     compileSdkVersion 27
-    buildToolsVersion "27.0.3"
 
     defaultConfig {
         minSdkVersion 21
         targetSdkVersion 27
```
[the remark this answers is not in the document]
Comment (lazar-mitrovic) [reply src/main/resources/native/android/android_project/app/build.gradle:8]: According to https://developer.android.com/studio/releases/build-tools :
```
If you're using Android plugin for Gradle 3.0.0 or higher, your project automatically uses a default version of the build tools that the plugin specifies.
```

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java [hunk 16]
```diff
@@ -183,7 +182,7 @@ public boolean runUntilEnd() throws IOException, InterruptedException {
                 "-H:+UseCAPCache",
                 "-H:CAPCacheDir=" + getCapCacheDir().toAbsolutePath().toString(),
                 "-H:+UseBionicC"));
-        if (useLLVM) {
+        if (projectConfiguration.isUseLLVM()) {
```
[the remark this answers is not in the document]
Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:193]: Should be fixed now

File: src/main/java/com/gluonhq/substrate/util/FileDeps.java (hunk 14 above)
Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/util/FileDeps.java:78]: It is known issue at Google's end https://issuetracker.google.com/issues/144111441#comment16
If package task doesn't fail I think we can safely ignore it.

Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/util/FileDeps.java:78]: Also, Gradle doesn't actually use NDK in this project configuration, since we are linking everything in `AbstractTarget` and just copying result as shared library.

File: src/main/resources/native/android/android_project/app/build.gradle [hunk 17]
```diff
@@ -12,4 +12,13 @@ android {
         implementation 'com.android.support:support-v4:27.1.1'
         api fileTree(dir: '../libs', include: '*.jar')
     }
+
+    signingConfigs {
+        release {
```
Comment (jperedadnr) [inline src/main/resources/native/android/android_project/app/build.gradle:17]: This is missing?
```
buildTypes {
        release {
            signingConfig signingConfigs.release
        }
    }
```

File: src/main/resources/native/android/android_project/app/build.gradle [hunk 18]
```diff
@@ -12,4 +12,13 @@ android {
...
+            storeFile file("KEYSTORE_FILE")
```
Comment (jperedadnr) [inline src/main/resources/native/android/android_project/app/build.gradle:18]: I'm not fond of adding the passwords to the file. I'd rather have them in a gradle.properties file, that can be global or local to the project.
like:
```
STOREPASS=real_store_password
...
```
These properties can be also set by the client maven plugin, like:
```
<providedKeyStorePassword>STOREPASS</providedKeyStorePassword >
...
```

File: src/main/resources/native/android/android_project/app/build.gradle (hunk 15 above)
[the remark this answers is not in the document]
Comment (jperedadnr) [reply src/main/resources/native/android/android_project/app/build.gradle:8]: I meant `targetSdkVersion`, we already set 28 in the manifest.

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (jperedadnr) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:401]: About that... `gensrc` should be just an intermediate folder for source files that might be modified by the developer if needed. Initially it was meant for AndroidManifest.xml or Default-info.plist, so these could be copied to `src/android` or `src/ios`. If the latter were present then these will be used.

My earlier comment was about keeping the manifest accessible, but the project could be still part of `gvm`, as it was before. 

I understand that now we have a more complex situation, as we have a whole gradle project, and it doesn't make sense that the developer moves the whole project to `src/android`. 

I'd say we need to keep in gensrc only the files that can be copied to src/android for further modifications: AndroidManifest, and maybe build.gradle?

File: src/main/resources/native/android/android_project/app/build.gradle (hunk 15 above)
[the remark this answers is not in the document]
Comment (lazanet) [reply src/main/resources/native/android/android_project/app/build.gradle:8]: According to this, newer buildTools are backwards compatible with older targetSdk's:
https://stackoverflow.com/a/24523113

Also `AndroidManifest.xml`' values should be removed since gradle ignores them anyway.

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (johanvos) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:401]: why would anyone want to modify those files manually?

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java [hunk 19]
```diff
@@ -223,20 +182,30 @@ public boolean runUntilEnd() throws IOException, InterruptedException {
...
     List<String> getTargetSpecificAOTCompileFlags() throws IOException {
-        return Arrays.asList("-H:CompilerBackend=" + Constants.BACKEND_LLVM,
+        ArrayList<String> flags = new ArrayList<String>(Arrays.asList(
                 "-H:-SpawnIsolates",
                 "-Dsvm.targetArch=" + projectConfiguration.getTargetTriplet().getArch(),
                 "-H:+UseOnlyWritableBootImageHeap",
                 "-H:+UseCAPCache",
                 "-H:CAPCacheDir=" + getCapCacheDir().toAbsolutePath().toString(),
-                "-H:CustomLD=" + ldlld.toAbsolutePath().toString());
+                "-H:+UseBionicC",
```
Comment (johanvos) [inline src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:191]: This fails on GraalVM 20.1
Can we make this optional until there is a GraalVM release that has this flag?

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java [hunk 20]
```diff
@@ -108,14 +94,16 @@ public AndroidTargetConfiguration( ProcessPaths paths, InternalProjectConfigurat
 
         Path clangguess = Paths.get(this.ndk, "toolchains", "llvm", "prebuilt", hostPlatformFolder, "bin", "clang");
         this.clang = Files.exists(clangguess) ? clangguess : null;
+
+        projectConfiguration.setBackend(Constants.BACKEND_AARCH64);
```
Comment (johanvos) [inline src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:98]: This won't work on 20.1, as it requires https://github.com/oracle/graal/commit/879a752eb9bae1aefa1576370414c668ebf0ec79

Comment (johanvos) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:98]: https://github.com/gluonhq/substrate/pull/615 should fix this.

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java (hunk 19 above)
Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:191]: Maybe we should add GraalVM version detection for this kind of things?

File: src/main/resources/native/android/android_project/app/build.gradle (hunk 18 above)
Comment (lazar-mitrovic) [reply src/main/resources/native/android/android_project/app/build.gradle]: Does this fix look ok to you?

File: src/main/resources/native/android/android_project/app/build.gradle (hunk 17 above)
Comment (lazar-mitrovic) [reply src/main/resources/native/android/android_project/app/build.gradle:17]: Should be fixed now

File: src/main/resources/native/android/android_project/app/build.gradle (hunk 15 above)
[the remark this answers is not in the document]
Comment (lazar-mitrovic) [reply src/main/resources/native/android/android_project/app/build.gradle]: Should be fixed

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:401]: Should be fixed now

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java [hunk 21]
```diff
@@ -407,95 +311,41 @@ private void copyOtherDalvikClasses() throws IOException, InterruptedException {
...
-        String zipAlignCmd = buildToolsPath.resolve("zipalign").toString();
-        ProcessRunner zipAlign = new ProcessRunner(zipAlignCmd, "-f", "4", unalignedApk, alignedApk);
-        return zipAlign.runProcess("zipalign");
-    }
+    /*
+     * Generates release signing configuration if
+     * keystore is provided
+     */
+    private void generateSigningConfiguration() throws IOException {
+        Path settingsFile = getAndroidProjectPath().resolve("keystore.properties");
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:347]: The file is not created, is it?

Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:353]: It is copied from here
https://github.com/gluonhq/substrate/blob/12fe324a2336c5cbea8aefb29ddcf834001e012a/src/main/resources/native/android/android_project/keystore.properties#L1-L4

Comment (jperedadnr) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:353]: I saw that, but what I meant is that the file is not created under `gensrc/android_project`. 

Also I don't see where we call release, all the tasks are `assembleDebug` or `installDebug`, and the generated apk is under `outputs/apk/debug/apk-debug`.

Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:353]: That file is copied to `/gvm/android_project/keystore.properties`. I don't think it should be in `gensrc`, since `keystore.properties` is auto-generated from `ReleaseConfiguration`, on each `package` task.

I don't call release since `./gradlew assembleRelease` is a lot slower, requires release keystore and isn't required for debugging. I believe that `client:publish` task or something similar might be better path moving forward.

Comment (jperedadnr) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:353]: Hmm, I still have `gensrc/android_project`, so I've must missed one of your latest commits (hard to say since you squash them ;) ). Will checkout again.

About `publish`, precisely we have added `releaseConfiguration` to the plugin: if you have it, you want to release, so `package` should do that.

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java [hunk 22]
```diff
@@ -108,14 +95,17 @@ public AndroidTargetConfiguration( ProcessPaths paths, InternalProjectConfigurat
 
         Path clangguess = Paths.get(this.ndk, "toolchains", "llvm", "prebuilt", hostPlatformFolder, "bin", "clang");
         this.clang = Files.exists(clangguess) ? clangguess : null;
+        if (projectConfiguration.getGraalVersion().compareTo(new Version("20.1.0")) > 0) {
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:98]: This fails for me, as I have installed `graalvm-ce-java11-20.2.0-dev`, and that produces ` GraalVM CE 20.2.0-dev`, that doesn't follow the current pattern (but this is a different issue though, I filed https://github.com/gluonhq/substrate/issues/617 for it).

Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:94]: Can you check if it works for you now?

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java (hunk 21 above)
Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:360]: I added check if `releaseConfiguration` is set to switch between Debug and Release

Comment (jperedadnr) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:360]: Ok, works now under gvm. However, I still think we should use `assembleRelease` if developer provides `releaseConfiguration` and calls `client:package`.

In order to test it, I'm using `assembleRelease`, and I got an exception:
```
Execution failed for task ':app:validateSigningRelease'.
> Keystore file not set for signing config release
```
I've done some changes:
- moved keystore.properties to `app`, and change the app's build to:

```
release {
            if (file("keystore.properties").exists()) {
                Properties keystoreSettings = new Properties()
                keystoreSettings.load(new FileInputStream(file("keystore.properties")))

                storeFile = file(keystoreSettings['storeFile'])
                storePassword = keystoreSettings['storePassword']
                keyAlias = keystoreSettings['keyAlias']
                keyPassword = keystoreSettings['keyPassword']
            }
        }

buildTypes {
        release {
            signingConfig signingConfigs.release
        }
    }

```

This works fine now (and it takes a few more seconds).

Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java]: I've added those changes in latest commit.
As for `client:package`, https://github.com/gluonhq/substrate/blob/bf585c51d5fbb9eaec6b1bfae86eb9e59c6bf90f/src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java#L367-L374 should be false if `releaseConfiguration` is provided, so "Release" should be returned and "assembleRelease" emitted.

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java [hunk 23]
```diff
@@ -127,23 +123,32 @@ public boolean packageApp() throws IOException, InterruptedException {
...
-                            "assembleDebug");
+                            "assemble" + configuration);
         assembleDebug.addToEnv("ANDROID_HOME", sdk);
         assembleDebug.addToEnv("JAVA_HOME", projectConfiguration.getGraalPath().toString());
-        return assembleDebug.runProcess("package-task") == 0;
+        if (assembleDebug.runProcess("package-task") != 0) {
+            return false;
+        }
+        Path generatedApk = getAndroidProjectPath().resolve("app").resolve("build")
+                            .resolve("outputs").resolve("apk").resolve(configuration.toLowerCase())
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:138]: Use `toLowerCase(Locale.ROOT)`.
Can you also check if `Files.exists(generatedApk)` before copying it to gvm? In case, for any reason, the process ends successfully but the apk file is not found?

Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java]: Should be fixed now

File: src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java [hunk 24]
```diff
@@ -89,7 +89,8 @@
             "-H:+AddAllCharsets",
             "-H:+ReportExceptionStackTraces",
             "-H:-DeadlockWatchdogExitOnTimeout",
-            "-H:DeadlockWatchdogInterval=0"
+            "-H:DeadlockWatchdogInterval=0",
+            "-H:+RemoveSaturatedTypeFlows"
```
Comment (johanvos) [inline src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java:93]: This would be better in a separate PR, as it affects all platforms.

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java [hunk 25]
```diff
@@ -532,86 +389,61 @@ private int sign(Path buildToolsPath, String alignedApk) throws IOException, Int
...
+     * Copies the Android project from the jar resource and stores it in
+     * a directory. Return that directory
+     */
+    private Path prepareAndroidProject() throws IOException {
+        Path androidProject = getAndroidProjectPath();
+        if (Files.exists(androidProject)) {
+            FileOps.deleteDirectory(androidProject);
         }
-        throw new BuildToolNotFoundException();
+        FileOps.copyDirectoryFromResources("/native/android/android_project", androidProject);
```
Comment (johanvos) [inline src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java:406]: `/native/android/android_project` can be a constant (error-prone to spell it)

**Comment by eugener** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gluonhq/substrate/pullRequest?prid=5574548&bid=18275932)

**Comment by eugener** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gluonhq/substrate/pullRequest?prid=5574548&bid=18275932)

File: src/main/java/com/gluonhq/substrate/Constants.java [hunk 26]
```diff
@@ -138,6 +139,7 @@
     public static final String DALVIK_PRECOMPILED_CLASSES = "/precompiled/classes/";
     public static final String DALVIK_ACTIVITY_PACKAGE = "com/gluonhq/helloandroid/";
     public static final String DALVIK_JAVAFX_PACKAGE = "javafx/scene/input/";
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/Constants.java:141]: Unused constants, first one can be used in `prepareAndroidManifest`. Second one can be removed

File: src/main/java/com/gluonhq/substrate/Constants.java [hunk 27]
```diff
@@ -138,6 +139,7 @@
...
+    public static final String DALVIK_PROJECT_PATH = "/native/android/android_project";
```
Comment (jperedadnr) [inline src/main/java/com/gluonhq/substrate/Constants.java:142]: Can you refactor the path `/native/android/` to one constant, and then only the name to another one? There are few uses for the former and also for the latter.

**Comment by eugener** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gluonhq/substrate/pullRequest?prid=5574548&bid=18275932)

File: src/main/java/com/gluonhq/substrate/Constants.java (hunk 27 above)
Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/Constants.java]: Fixed

File: src/main/java/com/gluonhq/substrate/Constants.java (hunk 26 above)
Comment (lazar-mitrovic) [reply src/main/java/com/gluonhq/substrate/Constants.java]: Manifest uses package name with dots as placeholder, so I removed this completely
